hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -43,13 +59,61 @@ public TokenRejectHandler() {}
     @Override
     public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
         requireNonNull(context);
-        // Todo implement the preHandle logic
+        final var txn = context.body();
+        final var op = txn.tokenRejectOrThrow();
+
+        // If owner account is specified, verify that it's valid and require that it signs the transaction.
+        // If not specified, payer account is considered as the one rejecting the tokens.
+        if (op.hasOwner()) {
+            final var accountStore = context.createStore(ReadableAccountStore.class);
+            verifySenderAndRequireKey(op.owner(), context, accountStore);
+        }
+    }
+
+    private void verifySenderAndRequireKey(
+            final AccountID senderId, final PreHandleContext context, final ReadableAccountStore accountStore)
+            throws PreCheckException {
+
+        final var senderAccount = accountStore.getAliasedAccountById(senderId);
+        validateTruePreCheck(senderAccount != null, INVALID_ACCOUNT_ID);
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = senderAccount.key();
+        if (key == null || !isValid(key)) {
+            throw new PreCheckException(ACCOUNT_IS_IMMUTABLE);
+        }
+        context.requireKey(key);
     }
 
+    @SuppressWarnings("java:S2259")
     @Override
     public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
-        // Todo: Implement the pureChecks logic
-        throw new UnsupportedOperationException("Not implemented yet.");
+        requireNonNull(txn, "Transaction body cannot be null");
+        final var op = txn.tokenRejectOrThrow();
+
+        validateFalsePreCheck(op.rejections().isEmpty(), INVALID_TRANSACTION_BODY);
+        if (op.hasOwner()) {
+            validateAccountID(op.owner(), null);
```

Comment:
Should this be `INVALID_OWNER_ID` or some other response code ?

Reply:
I've mentioned it in the [protobuf changes PR](https://github.com/hashgraph/hedera-protobufs/pull/351/files#r1599643110)